consensus/parlia/fusionfork.go

```diff
@@ -0,0 +1,214 @@
+
+	method := "getValidatorElectionInfo"
+	toAddress := common.HexToAddress(systemcontracts.StakeHubContract)
+	gas := (hexutil.Uint64)(uint64(math.MaxUint64 / 2))
+
+	data, err := p.stakeHubABI.Pack(method, big.NewInt(0), big.NewInt(0))
+	if err != nil {
+		log.Error("Unable to pack tx for getValidatorElectionInfo", "error", err)
+		return nil, nil, nil, err
+	}
+	msgData := (hexutil.Bytes)(data)
+
+	result, err := p.ethAPI.Call(ctx, ethapi.TransactionArgs{
+		Gas:  &gas,
+		To:   &toAddress,
+		Data: &msgData,
+	}, blockNr, nil, nil)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var totalLength *big.Int
+	if err := p.stakeHubABI.UnpackIntoInterface(&[]interface{}{&validators, &votingPowers, &voteAddrs, &totalLength}, method, result); err != nil {
+		return nil, nil, nil, err
+	}
+	if totalLength.Int64() != int64(len(validators)) {
+		return nil, nil, nil, fmt.Errorf("validator length not match")
+	}
+
+	return validators, votingPowers, voteAddrs, nil
+}
+
+func (p *Parlia) getMaxElectedValidators(blockNr rpc.BlockNumberOrHash) (maxElectedValidators *big.Int, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel() // cancel when we are finished consuming integers
+
+	method := "maxElectedValidators"
+	toAddress := common.HexToAddress(systemcontracts.StakeHubContract)
+	gas := (hexutil.Uint64)(uint64(math.MaxUint64 / 2))
+
+	data, err := p.stakeHubABI.Pack(method)
+	if err != nil {
+		log.Error("Unable to pack tx for maxElectedValidators", "error", err)
+		return nil, err
+	}
+	msgData := (hexutil.Bytes)(data)
+
+	result, err := p.ethAPI.Call(ctx, ethapi.TransactionArgs{
+		Gas:  &gas,
+		To:   &toAddress,
+		Data: &msgData,
+	}, blockNr, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := p.stakeHubABI.UnpackIntoInterface(&maxElectedValidators, method, result); err != nil {
+		return nil, err
+	}
+
+	return maxElectedValidators, nil
+}
+
+func getTopValidatorsByVotingPower(validators []common.Address, votingPowers []*big.Int, voteAddrs [][]byte, maxElectedValidators *big.Int) ([]common.Address, []uint64, [][]byte) {
+	var validatorHeap ValidatorHeap
+	for i := 0; i < len(validators); i++ {
+		// only keep validators with voting power > 0
+		if votingPowers[i].Cmp(big.NewInt(0)) == 1 {
+			validatorHeap = append(validatorHeap, ValidatorItem{
+				address:     validators[i],
+				votingPower: votingPowers[i],
+				voteAddress: voteAddrs[i],
+			})
+		}
+	}
+	hp := &validatorHeap
+	heap.Init(hp)
+
+	length := int(maxElectedValidators.Int64())
+	if length > len(validatorHeap) {
+		length = len(validatorHeap)
+	}
+	eValidators := make([]common.Address, length)
+	eVotingPowers := make([]uint64, length)
+	eVoteAddrs := make([][]byte, length)
+	for i := 0; i < length; i++ {
+		item := heap.Pop(hp).(ValidatorItem)
+		eValidators[i] = item.address
+		eVotingPowers[i] = new(big.Int).Div(item.votingPower, big.NewInt(1e10)).Uint64() // to avoid overflow
```

Comment:
Please add a comment in this file to explain why it is 1e10, otherwise it will be confused to reader.

Reply:
Done

Reply:
`BNB chain ` ==> `BNB Beacon Chain`

Reply:
Done